src/lib.rs

```diff
@@ -208,6 +229,122 @@ impl System {
         Ok(promise)
     }
 
+    fn unit_start_and_wait(mut cx: FunctionContext) -> JsResult<JsPromise> {
+        let rt = runtime(&mut cx)?;
+        let system = cx.argument::<JsBox<System>>(0)?;
+        let service_name = cx.argument::<JsString>(1)?.value(&mut cx);
+        let wait_interval = cx.argument::<JsNumber>(2)?.value(&mut cx) as u64;
+        let mode = cx.argument::<JsString>(3)?.value(&mut cx);
+        let channel = cx.channel();
+
+        let connection = system.connection.clone();
+        let (deferred, promise) = cx.promise();
+
+        // Start and wait functionality is defined in a separate function, which returns
+        // a Result, so that it is easier to propagate error conditions, e.g. unit does
+        // not exist, etc.
+        async fn start_and_wait_unit(
+            service_name: &str,
+            connection: Connection,
+            mode: &str,
+            wait_interval: u64,
+        ) -> anyhow::Result<(String, i32, String)> {
+            let unit_path_str = service_to_unit_path(service_name);
+
+            let unit_path =
+                OwnedObjectPath::try_from(unit_path_str.clone()).with_context(|| {
+                    format!("Cannot convert unit name `{service_name}` to service path")
+                })?;
+
+            let mut unit = UnitProxy::builder(&connection)
+                .path(unit_path.clone())
+                .with_context(|| format!("Cannot set unit path from {unit_path_str}"))?
+                .build()
+                .await
+                .with_context(|| format!("Cannot build unit proxy for {unit_path_str}"))?;
+
+            let mut active_state = unit
+                .active_state()
+                .await
+                .context("Failed to get unit active state")?;
+
+            // Create a D-Bus stream that watches for service active state changes
+            let mut stream = unit.receive_active_state_changed().await;
+
+            unit.start(mode)
+                .await
+                .with_context(|| format!("Failed to start unit {unit_path_str}"))?;
```

Comment:
The only thing that I'm unsure about is that with this implementation there is some disconnection between the current action we are taking and the state of the unit. This call may fail because there is a queued job already for the unit, or perhaps this is queued and the result we get on the active state is the result from a previous operation. Perhaps we could also monitor the Job state until it switches from `waiting` to `running`?

I guess it depends on the case, but for something like the OS update it  doesn't matter too much who started the update as long as you know that the process is running (even if the request came from a previous job) and have the exec status at the end.